pulsar/internal/connection.go

```diff
@@ -729,7 +756,7 @@ func (c *connection) Close() {
 
 func (c *connection) changeState(state connectionState) {
 	c.Lock()
-	c.state = state
+	atomic.StoreInt32(&c.state, int32(state))
```

Comment:
In here, Is it necessary for us to `c.Lock()` and `c.Unlock()` actions on an atomic operation?

Reply:
I think it's still needed to ensure that the mutex condition is broadcasted. To trigger a condition you need to have a lock on the associated mutex.

Reply:
I am wondering if there are other problems with this approach, because for an atomic primitive, we rarely see the operation of Lock and Unlock it.


```
atomic.StoreInt32(&c.state, int32(state))
c.Lock()
c.cond.Broadcast()
c.Unlock()
```

Reply:
There are 2 different usages here: 
 * On one hand we want other go routines to be able to check the state, without taking a lock
 * On the other hand, we still need to maintain the atomic state update and notification

Reply:
yes, i agree with you, The main difference between us here is whether we need to lock to protect the atomic primitive, I think it is not needed, atomic itself is a synchronization primitive, so here, we can reduce the granularity of the lock and only lock sync.cond. The code example is as follows:

```
atomic.StoreInt32(&c.state, int32(state))
c.Lock()
c.cond.Broadcast()
c.Unlock()
```

Reply:
Please ignore me